crates/sui-types/src/transaction.rs

```diff
@@ -1868,6 +1868,62 @@ pub struct SenderSignedTransaction {
 }
 
 impl SenderSignedData {
+    /// with cache
+    pub fn verify_message_signature_with_cache(
+        &self,
+        verify_params: &VerifyParams,
+        cache: &VerifiedDigestCache<ZKLoginInputsDigest>,
+    ) -> SuiResult {
+        fp_ensure!(
+            self.0.len() == 1,
+            SuiError::UserInputError {
+                error: UserInputError::Unsupported(
+                    "SenderSignedData must contain exactly one transaction".to_string()
+                )
+            }
+        );
+        if self.intent_message().value.is_system_tx() {
+            return Ok(());
+        }
+
+        // Verify signatures. Steps are ordered in asc complexity order to minimize abuse.
+        let signers = self.intent_message().value.signers();
+        // Signature number needs to match
+        fp_ensure!(
+            self.inner().tx_signatures.len() == signers.len(),
+            SuiError::SignerSignatureNumberMismatch {
+                actual: self.inner().tx_signatures.len(),
+                expected: signers.len()
+            }
+        );
+        // All required signers need to be sign.
+        let present_sigs = self.get_signer_sig_mapping()?;
+        for s in signers {
+            if !present_sigs.contains_key(&s) {
+                return Err(SuiError::SignerSignatureAbsent {
+                    signer: s.to_string(),
+                });
+            }
+        }
+
+        // Verify all present signatures.
+        for (signer, signature) in present_sigs {
+            match signature {
+                GenericSignature::ZkLoginAuthenticator(z) => {
+                    cache.is_verified(
+                        z.hash(),
+                        || signature.verify_claims(self.intent_message(), signer, verify_params),
+                        || z.verify_remaining_checks(self.intent_message(), signer, verify_params),
+                    )?;
```

Comment:
this is where the callback happened @mystenmark 

if the LRU never seen the proof, use `signature.verify_claims` to check everything, if it did, use `verify_remaining_checks` that only checks ephemeral sig.